docs/rfc/rfc-10.adoc

```diff
@@ -0,0 +1,248 @@
+the existing ones to address upgradeability concerns.
+
+```
+/// @notice Represents tBTC deposit request data.
+struct DepositRequest {
+    // (...)
+
+    // Optional extra data used by depositors who are smart contracts. The smart
+    // contract depositor can use the extra data to represent additional
+    // information, such as the depositor's address on a foreign (non-Ethereum)
+    // chain.
+    bytes32 depositorExtraData;
+}
+```
+
+To maintain backward compatibility with the existing deposit flow, we will
+implement another reveal function in the `Bridge` contract. The function will be
+mostly identical to the existing `revealDeposit` except that it will respect the
+alternative P2SH deposit script version allowing to pass 32 bytes of extra data.
+The extra data will not be stored in the Bridge. It is the depositor's
+responsibility to make the use of them.
+
+==== Solana gateway
+
+`SolanaGateway` contract will be deployed on Solana and should be placed in
+`tbtc-v2/cross-chain/solana` project. The contract will communicate with
+Ethereum L1 using Wormhole VAAs to reveal Bitcoin deposits. The contract will
+use Wormhole Token Bridge to request Bitcoin redemption on Ethereum L1.
+
+For the implementation, we should consider using the
+link:https://solang.readthedocs.io/en/latest/index.html[Solang compiler]. 
+
+The contract needs to be upgradeable.
+
+==== Solana depositor
+
+`SolanaDepositor` contract will act as tBTC depositor on L1 Ethereum and should
+be placed in `tbtc-v2/solidity` project. The contract will receive Wormhole VAAs
+from Solana and reveal Bitcoin deposits to the tBTC `Bridge` contract. The
+contract will map the revealed deposit to the Solana address. It will also
+bridge the received tBTC once the deposit is swept or tBTC is optimistically
+minted.
+
+This contract needs to have a reference to the `TBTCVault` to inspect the
+`optimisticMintingRequests` mapping and to the `Bridge` to inspect the
+`deposits` mapping to confirm the state when it is notified tBTC has been minted
+for the given deposit. 
+
+The contract needs to be upgradeable.
+
+==== Solana redeemer
+
+`SolanaRedeemer` contract will act as tBTC redeemer on L1 Ethereum and should be
+placed in `tbtc-v2/solidity` project. The contract will receive tBTC from Solana
+via Wormhole Token Bridge and request redemption in the tBTC `Bridge` contract. 
+
+The contract needs to be upgradeable.
+
+==== Relayer bot
+
+To optimize the user experience, a relayer bot needs to be implemented. The
+relayer's responsibility will be:
+
+- Deliver Wormhole VAA to `SolanaDepositor` contract once the deposit was
+  revealed on Solana.
+- Request bridging tBTC from Ethereum to Solana once the tBTC for the deposit
+  has been optimistically minted or the deposit was swept.
+- Redeem bridged tBTC on Solana to the depositor address once the tBTC minted on
+  L1 has been successfully bridged to Solana.
```

Comment:
I don't follow this point, can you check it's valid?

Shouldn't it be:
```suggestion
- Transfer bridged tBTC on Solana to the depositor address once the tBTC minted on
  L1 has been successfully bridged to Solana.
```

Reply:
I think the original version is more precise. This RFC is pretty heavily oriented around Wormhole bridge and "redeem" is what is happening in this step in Wormhole, see https://www.portalbridge.com/#/redeem.

tl;dr; is that when you bridge assets, you *deposit* them in the Wormhole Token Bridge contract on the particular source chain. Then, Wormhole guardians produce VAA and you need to take that VAA and *redeem* the wrapped token on the target chain using it.

Reply:
I tried to disambiguate them in b1d234e98ffcc82c3a8e73257d3d8312eff70c6a.